pkg/ingester/ingester.go

```diff
@@ -2893,33 +2956,34 @@ func (i *Ingester) shipBlocks(ctx context.Context, allowed *util.AllowedTenants)
 	})
 }
 
-func (i *Ingester) compactionLoop(ctx context.Context) error {
+// compactionServiceRunning is the running function of internal service responsible to periodically
+// compact TSDB Head.
+func (i *Ingester) compactionServiceRunning(ctx context.Context) error {
 	// At ingester startup, spread the first compaction over the configured compaction
 	// interval. Then, the next compactions will happen at a regular interval. This logic
 	// helps to have different ingesters running the compaction at a different time,
 	// effectively spreading the compactions over the configured interval.
-	firstHeadCompactionInterval := i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval
-	if i.cfg.BlocksStorageConfig.TSDB.HeadCompactionIntervalJitterEnabled {
-		firstHeadCompactionInterval = util.DurationWithNegativeJitter(i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval, 1)
-	}
-
-	ticker := time.NewTicker(firstHeadCompactionInterval)
-	defer ticker.Stop()
-	tickerRunOnce := false
+	firstInterval, standardInterval := i.compactionServiceInterval()
+	stopTicker, tickerChan := util.NewVariableTicker(firstInterval, standardInterval)
+	defer stopTicker()
 
 	for ctx.Err() == nil {
 		select {
-		case <-ticker.C:
+		case <-tickerChan:
 			// The forcedCompactionMaxTime has no meaning because force=false.
 			i.compactBlocks(ctx, false, 0, nil)
 
 			// Check if any TSDB Head should be compacted to reduce the number of in-memory series.
 			i.compactBlocksToReduceInMemorySeries(ctx, time.Now())
```

Comment:
Should we be doing early compactions during ingest-reader catch-up? Unless catch-up takes more than 20 minutes, all series ingested via ingest-reader catch up will be active series.

Reply:
It's true that we don't expect to be used, but it shouldn't hurt keeping it. The reason why it doesn't hurt is because it can't continuously trigger head compaction because of `-blocks-storage.tsdb.early-head-compaction-min-estimated-series-reduction-percentage` (default is 15%). This settings means that the early head compaction is triggered only if it would reduce at least 15% of in-memory series in ingester. If not, it will not do the early compaction at all.